packages/core-plugin-api/src/extendable-components/ExtensionsProvider.tsx

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useContext, PropsWithChildren, useMemo } from 'react';
+import {
+  createVersionedValueMap,
+  createVersionedContext,
+} from '@backstage/version-bridge';
+
+import {
+  ComponentExtension,
+  ExtendableComponentPropsInterceptor,
+  ExtendableComponentProvider,
+  ExtendableComponentRef,
+} from './types';
+
+/**
+ * Prop types for the ApiProvider component.
+ * @public
+ */
+export type ExtensionsProviderProps = {
+  /**
+   * An extension or an array of extensions
+   */
+  extensions: ComponentExtension<any, any>[] | ComponentExtension<any, any>;
+};
+
+type ExtensionMapKey = ExtendableComponentRef<any, any, any>;
```

Comment:
The component ref objects aren't guaranteed to be unique of any given extendable component. You can and most likely will have duplicate installations of for example `@backstage/core-components`. I think the best approach is most likely to add a string ID for each component ref and use that as the key instead.

Reply:
Is that how api refs are handled internally? If so, yes it makes a lot of sense.

Reply:
Yep, they're keyed by the ID